Source/Parser/Functions/ArrayReduceFunction.cs

```diff
@@ -0,0 +1,83 @@
+﻿using RATools.Parser.Expressions;
+using System;
+using System.Linq;
+
+namespace RATools.Parser.Functions
+{
+    internal class ArrayReduceFunction : FunctionDefinitionExpression
+    {
+        public ArrayReduceFunction()
+            : base("array_reduce")
+        {
+            Parameters.Add(new VariableDefinitionExpression("inputs"));
+            Parameters.Add(new VariableDefinitionExpression("initial"));
+            Parameters.Add(new VariableDefinitionExpression("reducer"));
+        }
+
+        public override bool Evaluate(InterpreterScope scope, out ExpressionBase result)
+        {
+            var array = GetParameter(scope, "inputs", out result);
+            if (array == null)
+            {
+                return false;
+            }
+            if (!array.ReplaceVariables(scope, out result))
+            {
+                return false;
+            }
+            var inputs = result as IIterableExpression;
+            if (inputs == null)
+            {
+                result = new ErrorExpression("Cannot iterate over " + result.ToString(), result);
+                return false;
+            }
+
+            var accumulatorExpression = GetParameter(scope, "initial", out result);
+            if (accumulatorExpression == null)
+            {
+                return false;
+            }
+
+            var funcParam = GetFunctionParameter(scope, "reducer", out result);
+            if (funcParam == null)
+            {
+                return false;
+            }
+            if ((funcParam.Parameters.Count - funcParam.DefaultParameters.Count) != 2)
+            {
+                result = new ErrorExpression("reducer function must accept two parameters (acc, value)");
+                return false;
+            }
+
+            var iteratorScope = funcParam.CreateCaptureScope(scope);
+            foreach (var kvp in funcParam.DefaultParameters)
+            {
+                iteratorScope.AssignVariable(new VariableExpression(kvp.Key), kvp.Value);
+            }
+
+            foreach (var input in inputs.IterableExpressions())
+            {
+                if (!input.ReplaceVariables(iteratorScope, out result))
+                {
+                    return false;
+                }
+                iteratorScope.AssignVariable(new VariableExpression(funcParam.Parameters.First().Name), accumulatorExpression);
+                iteratorScope.AssignVariable(new VariableExpression(funcParam.Parameters.Last().Name), input);

```

Comment:
This should be `funcParam.Parameters.ElementAt(1).Name`, as default parameters might appear after it.

Also recommend capturing the parameter names outside the loop. You could construct and reuse two variable expressions for all iterations.

Reply:
I believe you want to assign `result` to the second parameter, not `input`.